lib/Lookup.php

```
@@ -0,0 +1,147 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+
+namespace OCA\GlobalSiteSelector;
+
+
+use OCP\Federation\ICloudIdManager;
+use OCP\Http\Client\IClientService;
+use OCP\IConfig;
+use OCP\ILogger;
+
+/**
+ * Class Lookup
+ *
+ * query the lookup server to find the users location
+ *
+ * @package OCA\GlobalSiteSelector
+ */
+class Lookup {
+
+	/** @var IClientService */
+	private $httpClientService;
+
+	/** @var  string */
+	private $lookupServerUrl;
+
+	/** @var ILogger */
+	private $logger;
+
+	/** @var  ICloudIdManager */
+	private $cloudIdManager;
+
+	/**
+	 * Lookup constructor.
+	 *
+	 * @param IClientService $clientService
+	 * @param IConfig $config
+	 * @param ILogger $logger
+	 * @param ICloudIdManager $cloudIdManager
+	 */
+	public function __construct(IClientService $clientService,
+								IConfig $config,
+								ILogger $logger,
+								ICloudIdManager $cloudIdManager
+	) {
+		$this->httpClientService = $clientService;
+		$this->lookupServerUrl = $config->getSystemValue('lookup_server', '');
+		$this->logger = $logger;
+		$this->cloudIdManager = $cloudIdManager;
+	}
+
+	/**
+	 * try to find the exact user at the lookup server, we allow to search for
+	 * email addresses and federated cloud ids and internal UIDs.
+	 *
+	 * @param string $uid
+	 * @return string the url of the server where the user is located
+	 */
+	public function search($uid) {
+
+		$result = '';
+
+		// admin need to specify a lookup server with GSS capabilities
+		if (empty($this->lookupServerUrl)) {
+			$this->logger->error(
+				'Can not lookup user, no lookup server registered',
+				['app' => 'globalsiteselector']
+			);
+			return $result;
+		}
+
+		try {
+			$body = $this->queryLookupServer($uid);
+
+			if (isset($body['federationId'])) {
+				$result = $this->getUserLocation($body['federationId']);
+			}
+
+		} catch (\Exception $e) {
```

Never just eat exceptions. I would log them. Let me add this.